subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/modulecache/ModuleMetadataStore.java

```diff
@@ -22,35 +22,46 @@
 import org.gradle.internal.UncheckedException;
 import org.gradle.internal.component.external.model.ModuleComponentResolveMetadata;
 import org.gradle.internal.component.external.model.MutableModuleComponentResolveMetadata;
+import org.gradle.internal.resource.local.DefaultPathKeyFileStore;
+import org.gradle.internal.resource.local.FileAccessTracker;
 import org.gradle.internal.resource.local.LocallyAvailableResource;
 import org.gradle.internal.resource.local.PathKeyFileStore;
+import org.gradle.internal.resource.local.TouchingFileAccessTracker;
 import org.gradle.internal.serialize.kryo.KryoBackedDecoder;
 import org.gradle.internal.serialize.kryo.KryoBackedEncoder;
 
 import java.io.File;
 import java.io.FileInputStream;
 import java.io.FileOutputStream;
 
+import static java.util.Collections.singleton;
+
 public class ModuleMetadataStore {
```

Comment:
This store is wrapped in an in-memory cache, so `markAccessed()` will not be called for a module cached in-memory. I guess this means that a frequently used module could appear to be stale when a long-lived daemon retains it in the in-memory cache. I don't suppose this is a big problem, but something to be aware of.

I would advocate for not cleaning up the binary module descriptors, to avoid this issue (and other potential ones).